core/src/main/scala/cats/data/NonEmptyList.scala

```diff
@@ -286,6 +286,22 @@ final case class NonEmptyList[+A](head: A, tail: List[A]) extends NonEmptyCollec
     go(head, tail, Nil)
   }
 
+  /**
+   * Zips this `NonEmptyList` with another `NonEmptyList` and returns the pairs of elements.
+   *
+   * {{{
+   * scala> import cats.data.NonEmptyList
+   * scala> val as = NonEmptyList.of(1, 2, 3)
+   * scala> val bs = NonEmptyList.of("A", "B", "C")
+   * scala> as.zip(bs)
+   * res0: cats.data.NonEmptyList[(Int, String)] = NonEmptyList((1,A), (2,B), (3,C))
+   * }}}
+   */
+  def zip[B](nel: NonEmptyList[B]): NonEmptyList[(A, B)] = {
+    val (NonEmptyList(head1, tail1), NonEmptyList(head2, tail2)) = (this, nel)
```

Comment:
This relies on the compiler to avoid a needless allocation. Can we just explicitly write the code without the pattern matching or tuple?

```scala
NonEmptyList((head, nel.head), tail.zip(nel.tail))
```

Reply:
Good suggestions. Applied it in https://github.com/typelevel/cats/pull/3730/commits/d5605656277dcfa14c537a328d3f51e0e3d7a357!